src/Service/MailchimpSyncContact.php

```diff
@@ -0,0 +1,166 @@
+                ]
+            );
+        }
+    }
+
+    public function subscribe(User $user)
+    {
+        $this->requestApi(
+            $user,
+            'PUT',
+            $this->container->getParameter('mailchimp.api_url').'/'.md5(strtolower($user->getEmail())),
+            ['status' => self::STATUS_SUBSCRIBED]
+        );
+    }
+
+    public function unsubscribe(User $user)
+    {
+        $this->requestApi(
+            $user,
+            'PUT',
+            $this->container->getParameter('mailchimp.api_url').'/'.md5(strtolower($user->getEmail())),
+            ['status' => self::STATUS_UNSUBSCRIBED]
+        );
+    }
+
+    public function checkSubscriptionStatus(User $user)
+    {
+        return $this->requestApi(
+            $user,
+            'GET',
+            $this->container->getParameter('mailchimp.api_url').'/'.md5(strtolower($user->getEmail()))
+        );
+    }
+
+    private function requestApi(
+        User $user,
+        string $method,
+        string $url,
+        array $data = null,
+        bool $isNewContact = false
+    ) {
+        $options = [];
+        $expectedStatus = null;
+        if ($data) {
+            $options = ['body' => json_encode($data)];
+            if ($data['status']) {
+                $expectedStatus = $isNewContact ? self::STATUS_ADDED : $data['status'];
+            }
+        }
+
+        try {
+            $jsonContent = $this->httpClient->request(
+                $method,
+                $url,
+                $options
+            )
+                ->getContent();
+
+            $content = json_decode($jsonContent);
+            if (!empty($content) && !empty($content['status'])) {
+                return $content['status'];
+            }
+        } catch (TransportExceptionInterface | ClientExceptionInterface | RedirectionExceptionInterface | ServerExceptionInterface $e) {
+            $this->logger->error($e);
+        }
+
+        if ($expectedStatus) {
+            $this->manageSyncFail($user, $expectedStatus);
+        }
+
+        return null;
+    }
+
+    private function manageSyncFail(User $user, string $type)
+    {
+        $isSubscription = self::STATUS_UNSUBSCRIBED !== $type;
+        $action = $isSubscription ? 'inscrire à' : 'désincrire de';
+
+        $this->flashBag->add(
+            'error',
+            'Nous ne sommes pas parvenus à vous '.$action.' notre lettre d’actualités.<br>'.
+            'Un message a été envoyé à l’administrateur du site afin de régler le problème'
+        );
+
+        $mailMessage = $this->twig->render('emails/newsletter-sync.html.twig', [
+            'user' => $user,
+            'type' => $type,
+        ]);
+
+        $this->mailer->send(
+            $user->getEmail(),
+            //'contact@obs-saisons.fr',
+            'idir.alliche.tb@gmail.com',
+            $this->mailer->getSubjectFromTitle($mailMessage),
+            $mailMessage
+        );
+
+        $user->setIsNewsletterSubscriber($isSubscription);
```

Comment:
Why do you set that here?

Reply:
oh sorry I was meaning `$user->setIsNewsletterSubscriber(!$isSubscription);`

Reply:
if subscription or unsubscription didn't work you may want to roll it back in database too